crypt.go

```diff
@@ -237,6 +238,132 @@ func getPassword(sourceDevicePath, description string, reader io.Reader) (string
 	return askPassword(sourceDevicePath, "Please enter the "+description+" for disk "+sourceDevicePath+":")
 }
 
+// SnapModelChecker is used for verifying whether a Snap device model is
+// authorized to access the data on a volume unlocked by this package.
+type SnapModelChecker struct {
+	volumeName string
+	keyData    *KeyData
+	auxKey     AuxiliaryKey
+}
+
+// IsModelAuthorized indicates whether the supplied Snap device model is
+// authorized to access the data on the decrypted volume with the device
+// mapper name returned by VolumeName.
+func (c *SnapModelChecker) IsModelAuthorized(model SnapModel) (bool, error) {
+	return c.keyData.IsSnapModelAuthorized(c.auxKey, model)
+}
+
+// VolumeName is the device mapper name of the volume associated with this
+// SnapModelChecker.
+func (c *SnapModelChecker) VolumeName() string {
+	return c.volumeName
+}
+
+type activateWithKeyDataError struct {
+	k   *KeyData
+	err error
+}
+
+func (e *activateWithKeyDataError) Error() string {
+	return fmt.Sprintf("%s: %v", e.k.ID(), e.err)
+}
+
+func (e *activateWithKeyDataError) Unwrap() error {
+	return e.err
+}
+
+type keyDataAndError struct {
+	*KeyData
+	err error
+}
+
+type activateWithKeyDataState struct {
+	volumeName       string
+	sourceDevicePath string
+	activateOptions  []string
+	keyringPrefix    string
+
+	keys []*keyDataAndError
+
+	keyData *KeyData
+	auxKey  AuxiliaryKey
+}
+
+func (s *activateWithKeyDataState) errors() (out []*activateWithKeyDataError) {
+	for _, k := range s.keys {
+		if k.err == nil {
+			continue
+		}
+		out = append(out, &activateWithKeyDataError{k: k.KeyData, err: k.err})
+	}
+	return
+}
+
+func (s *activateWithKeyDataState) snapModelChecker() *SnapModelChecker {
+	return &SnapModelChecker{s.volumeName, s.keyData, s.auxKey}
```

Comment:
what's the plan for TPM going forward about this, to do this and also measure the model, to do only this, that means changing what we measure, also current key locking is based on the PCR used for the model? what will be the relation (if any) of  auxKey vs the reseal auth key?

Reply:
When using the KeyData API, all platforms will use SnapModelChecker, but there's no need to change what we measure to the TPM in the TPM case. Measuring the model there has uses beyond full-disk encryption anyway (eg, providing attestations).

In the TPM case, auxKey will be used to derive the elliptic key used for signing new PCR policies (as opposed to providing the key directly). The current TPMPolicyAuthKey type will go away and be replaced with the new type, but the code can support both old and new keys in order to aid migration - it's easy to determine whether auxKey is actually the elliptic key by deriving the public point from it and comparing it to the stored public key.

Reply:
that is reasonable but are there enough bits in aux keys for the double usage to derive both the EC key and the to derive a hmac secret for models?

Reply:
The derived keys protect the same resource, so the 32-bits in the "seed" is sufficient to derive both keys.

Reply:
Err, 32 *bytes* :)